Task: Automatically update CodeMeta upon release #1276

OlawumiSalaam opened the pull request:
Thank you for taking your time to contribute to Ersilia, just a few checks before we proceed
- [x] Have you followed the guidelines in our [Contribution Guide](https://github.com/ersilia-os/ersilia/blob/master/CONTRIBUTING.md)
- [x] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

**Description**

This task is to update the existing GitHub Actions workflow for the Ersilia project by automating the update of the version number in the codemeta.json file during the monthly release process. The existing workflow already updates the version in CITATION.cff, and this task ensures that the same version number is also updated in codemeta.json for consistency across project metadata files.
Ersilia has an automatic release trigger on the first day of each month at 3:00 AM (UTC). The workflow manages the following:

- Incrementing the version number in the pyproject.toml file.
- Creating and pushing a new release tag to the repository.
- Updating the version number in CITATION.cff.
- Publishing the release to GitHub, PyPI, Conda, and DockerHub.

This task adds a new step to ensure that the version in codemeta.json is updated to match the incremented version number.


**Changes to be made**

- A new step was introduced into the existing GitHub Actions workflow to update the version number in the codemeta.json file using the extracted version number. This ensures that the version field in this file remains consistent with the version number used for releases.
- The version number is updated programmatically using jq, a command-line JSON processor used to manipulate the JSON file. The jq command reads the existing codemeta.json file, updates the version field with the new version number stored in the VERSION environment variable, and writes the updated content back to codemeta.json.
- The CITATION.cff and codemeta.json files are committed and pushed to the repository in a single step, which ensures that the version is updated in both files. This ensures the files are updated before proceeding with the rest of the release process.


**Status**

- A new step introduced in the workflow to automatically update the version in the CodeMeta file.
- Pushed changes to the feature/update-codemeta-upon-release branch for review.


**To do**

- Run tests locally to verify the functionality. 
-when i tried to run locally, I encountered error: 
"**fatal:** could not read Username for 'https://github.com/': No such device or address,"


Is this pull request related to any open issue? If yes, replace issueID below with the issue ID
Yes,
Closes #1276

P-Mefut commented:
Hi @OlawumiSalaam nice but you haven't made changes to update  the URL yet when changes to the version are done

OlawumiSalaam commented:
@P-Mefut Thank you for your feedback. I will check again and implement.

DhanshreeA commented on .github/workflows/publish.yaml:
```diff
@@ -33,12 +32,10 @@ jobs:
           git config --local user.name "ersilia-bot"
           git tag -a "v${{ steps.version.outputs.VERSION }}" -m "v${{ steps.version.outputs.VERSION }}"
-          # git push "${repo}" origin "v${{ steps.version.outputs.VERSION }}"
+          git push origin "v${{ steps.version.outputs.VERSION }}"
```
I think we can remove this line entirely. I added it when I was trying to push tags through a script but that happens implicitly when we set `tags` to true in the commit and push action.

OlawumiSalaam replied:
There is no explicit git push for tags again. It is being handled implicitly when `tags: true ` is set in the commit and push action.

DhanshreeA commented on .github/workflows/publish.yaml:
```diff
@@ -57,12 +54,23 @@ jobs:
           author_name: "ersilia-bot"
           author_email: "ersilia-bot@users.noreply.github.com"
-          message: "update version [skip ci]"
+          message: "Update version in CITATION.cff and codemeta.json [skip ci]"
```
I guess we can just keep it as `Update version` because this step is updating a bunch of different files: `_static_version.py` in ersilia, `pyproject.toml`, the `CITATION.cff` file, and thanks to your work, the `codemeta.json` as well. We can add a descriptive comment in the beginning of the workflow to explain which files get updated in this pipeline.

OlawumiSalaam replied:
The commit message now uses "Update version"

OlawumiSalaam replied:
I added this comment at the start of the workflow to explain the the files being updated in the pipeline.
 _This job updates _static_version.py, pyproject.toml, CITATION.cff, and codemeta.json_

DhanshreeA commented on .github/workflows/publish.yaml:
```diff
@@ -1,25 +1,24 @@
+    - cron: '0 3 1 * *' # run at 3:00 AM on the first day of the month
+
 jobs:
   version:
```
We can add a comment here about the files updated as referenced in a later comment in this review.

OlawumiSalaam replied:
I have added a comment next to the cron job

DhanshreeA commented on .github/workflows/publish.yaml:
```diff
@@ -80,29 +88,25 @@ jobs:
           year=$(date +'%Y')
           title="$month $year"
-          version=`sed -n 's/.*version = "\([^"]*\)".*/\1/p' _static_version.py`
+          version=${{ needs.version.outputs.VERSION }}
```
Are we sure this will work? I am not very sure with passing variables between jobs in a GH actions pipeline, and I don't want this failing next month so I'll just leave this as is. :/

OlawumiSalaam replied:
I have left the `sed` command for version extraction from `_static_version.py` unchanged you suggested. But here is my recommendation:
 In the version job, we are extracting the version using the static_version_writer.py script and we are also using `sed` to download the `_static_version.py`  again. This can cause redundancy, we can consider using the version extracted in the `version` job and pass it as an output throughout the pipeline.

DhanshreeA commented:
LGTM generally @OlawumiSalaam! I really like how you took the time to clean up some of the existing code. :)

OlawumiSalaam commented:
@DhanshreeA thank you for the feedback. I will incorporate your feedback. Thank you.

OlawumiSalaam commented:
@DhanshreeA I have reviewed and implemented your feedback through the following steps:
I removed this line from the workflow:
`git push "${repo}" origin "v${{ steps.version.outputs.VERSION }}"`
I updated the message to reflect there are several files being updated:
"update message"
I added a comment to explain the files that are being updated by the cron job:
 _This job updates _static_version.py, pyproject.toml, CITATION.cff, and codemeta.json_
I kept the line unchanged for now and stated my observations:
`version=`sed -n 's/.*version = "\([^"]*\)".*/\1/p' _static_version.py`
`
observations:
In the version job, we are extracting the version using the static_version_writer.py script and we are also using sed to download the _static_version.py again. This can cause redundancy, we can consider using the version extracted in the version job and pass it as an output throughout the pipeline.

OlawumiSalaam commented:
@DhanshreeA Please review and I awaits your further feedback. Thank you.

OlawumiSalaam commented:
@DhanshreeA please I am waiting for your feedback on the updated task. Thank you

DhanshreeA commented on .github/workflows/publish.yaml:
```diff
@@ -57,12 +56,23 @@ jobs:
           author_name: "ersilia-bot"
           author_email: "ersilia-bot@users.noreply.github.com"
-          message: "update version [skip ci]"
+          message: "Update version"
```
@OlawumiSalaam we do need to keep `[skip ci]` in the commit message.

OlawumiSalaam commented:
@DhanshreeA Thank you for your review and I have implemented your feedback. Can I work on another issue

DhanshreeA commented:
> @DhanshreeA Thank you for your review and I have implemented your feedback. Can I work on another issue

Yes please go ahead, thank you!

DhanshreeA commented:
> Hi @OlawumiSalaam nice but you haven't made changes to update the URL yet when changes to the version are done

Good catch @P-Mefut!
